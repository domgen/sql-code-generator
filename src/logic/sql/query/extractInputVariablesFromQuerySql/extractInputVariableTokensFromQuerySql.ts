const EXTRACT_INPUT_VARIABLES_REGEX = /(?:[^\w])(\:\w+)/g;

export const extractInputVariableTokensFromQuerySql = ({ sql }: { sql: string }) => {
  const regex = new RegExp(EXTRACT_INPUT_VARIABLES_REGEX);
  const regexMatches = sql.match(regex) ?? [];
  const inputVariableTokens = regexMatches.map((str) => str.slice(1)); // drop the leading non-word character (TODO: find how to "match" without including that char)
  return inputVariableTokens;
};
