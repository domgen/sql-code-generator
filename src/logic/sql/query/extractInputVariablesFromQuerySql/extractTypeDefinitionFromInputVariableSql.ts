import { TypeDefinitionOfQueryInputVariable } from '../../../../model/valueObjects/TypeDefinitionOfQueryInputVariable';
import { throwErrorIfTableReferencePathImpliesTable } from '../common/throwErrorIfTableReferencePathImpliesTable';

export const extractTypeDefinitionFromInputVariableSql = ({ token, sql }: { token: string; sql: string }) => {
  // 1. check if this token matches the "resource.column = :token" pattern; if so, then the input type = the resource.column type
  const [
    _, // tslint:disable-line no-unused
    leftEqualsTableReferencePath, // check if "resource.column = :token"
  ] = new RegExp(`(\\w+\\.?\\w*)\\s?=\\s?(?:${token})`, 'g').exec(sql) ?? [];
  if (leftEqualsTableReferencePath) {
    throwErrorIfTableReferencePathImpliesTable({ referencePath: leftEqualsTableReferencePath });
    return new TypeDefinitionOfQueryInputVariable({
      name: token.replace(':', ''),
      tableReferencePath: leftEqualsTableReferencePath,
      functionReferencePath: null,
    });
  }

  // 2. check if this token matches the ":token = resource.column" pattern; same as above, but flipped
  const [
    __, // tslint:disable-line no-unused
    rightEqualsTableReferencePath, // check if "resource.column = :token"
  ] = new RegExp(`(?:${token})\\s?=\\s?(\\w+\\.?\\w*)`).exec(sql) ?? [];
  if (rightEqualsTableReferencePath) {
    throwErrorIfTableReferencePathImpliesTable({ referencePath: rightEqualsTableReferencePath });
    return new TypeDefinitionOfQueryInputVariable({
      name: token.replace(':', ''),
      tableReferencePath: rightEqualsTableReferencePath,
      functionReferencePath: null,
    });
  }

  // 3. check if this token is used in a function. If so, its equivalent to whatever is at that index of the function params

  // note: this reg matches the whole function def (e.g., `upsert_image(:url,:caption,:credit)`)
  const reg = `\\s+(\\w+\\((?:\\s*[:\\w]+,)*(?:\\s*${token},?)(?:\\s*[:\\w]+,?)*\\s?\\))`;
  const [
    ___, // tslint:disable-line no-unused
    tokenInsideFunctionMatch, // check if "resource.column = :token"
  ] = new RegExp(reg).exec(sql) ?? [];
  if (tokenInsideFunctionMatch) {
    // 1. grab the function name
    const functionName = tokenInsideFunctionMatch.split('(')[0].trim();

    // 2. figure out where in the array the token is
    const sqlAfterOpen = tokenInsideFunctionMatch.split('(')[1];
    const sqlBetweenOpenAndClose = sqlAfterOpen.split(')')[0];
    const parametersArray = sqlBetweenOpenAndClose.split(',').map((str) => str.trim());
    const index = parametersArray.indexOf(token);
    if (index === -1) throw new Error('unexpected'); // this should never occur if the regex is working correctly; fail fast though

    // return the function path
    return new TypeDefinitionOfQueryInputVariable({
      name: token.replace(':', ''),
      functionReferencePath: `${functionName}.${index}`,
      tableReferencePath: null,
    });
  }

  // if none of the above worked, throw an error. one should work...
  throw new Error(`could not extract type definition for input variable ${token}`);
};
