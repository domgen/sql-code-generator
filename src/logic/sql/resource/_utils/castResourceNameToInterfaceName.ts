import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../../model';

export const castResourceNameToInterfaceName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
