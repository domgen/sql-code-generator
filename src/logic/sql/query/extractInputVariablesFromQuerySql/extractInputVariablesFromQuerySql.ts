/*
  goal:
  - try to define input type when possible
  - fallback to "any" when not possible and ask user to add an issue w/ use case when this occurs
    - so that the generator is non-blocking

  e.g.,:
    - where u.id = :id => "{ id: number }"
    - where lower(u.name) = :name => { name: any } (for mvp);
      - later, { name: string } (since we'd track standard fn's; TODO)

  steps:
    - 1. find all of the `:__variable_name__` pattern input definitions (yesql format)
    - 2. attempt to figure out the type for each variable name
*/

import { extractInputVariableTokensFromQuerySql } from './extractInputVariableTokensFromQuerySql';
import { extractTypeDefinitionFromInputVariableSql } from './extractTypeDefinitionFromInputVariableSql';

export const extractInputVariablesFromQuerySql = ({ sql }: { sql: string }) => {
  // 1. find each input variable that needs to have a type def defined
  const inputVariableTokens = extractInputVariableTokensFromQuerySql({ sql });

  // 2. define the type definition for each variable
  const definitions = inputVariableTokens.map((token) => extractTypeDefinitionFromInputVariableSql({ token, sql }));

  // 3. return defs
  return definitions;
};
