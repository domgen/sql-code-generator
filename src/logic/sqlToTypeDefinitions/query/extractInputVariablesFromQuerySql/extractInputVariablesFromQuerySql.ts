import strip from 'sql-strip-comments';

import { TypeDefinitionOfQueryInputVariable } from '../../../../domain/objects/TypeDefinitionOfQueryInputVariable';
import { flattenSqlByReferencingAndTokenizingSubqueries } from '../common/flattenSqlByReferencingAndTokenizingSubqueries';
import { extractInputVariableTokensFromQuerySql } from './extractInputVariableTokensFromQuerySql';
import { extractTypeDefinitionFromInputVariableSql } from './extractTypeDefinitionFromInputVariableSql';

/**
 * goal:
 *  - try to define input type when possible
 * - fallback to "any" when not possible and ask user to add an issue w/ use case when this occurs
 *   - so that the generator is non-blocking
 *
 *  e.g.,:
 *   - where u.id = :id => "{ id: number }"
 *   - where lower(u.name) = :name => { name: any } (for mvp);
 *     - later, { name: string } (since we'd track standard fn's; TODO)
 *
 * steps:
 * - 1. find all of the `:__variable_name__` pattern input definitions (yesql format)
 * - 2. attempt to figure out the type for each variable name
 */

export const extractInputVariablesFromQuerySql = ({ sql }: { sql: string }) => {
  // find each input variable that needs to have a type def defined
  const inputVariableTokens = extractInputVariableTokensFromQuerySql({ sql });

  // suffix each token with the order in which it was found, since we may have the same token identified more than once
  const tokenToSuffixedTokensMap: { [index: string]: string[] } = {};
  const suffixedInputVariableTokens = inputVariableTokens.map((token) => {
    if (!tokenToSuffixedTokensMap[token]) tokenToSuffixedTokensMap[token] = [];
    const tokenOccurrenceIndex = tokenToSuffixedTokensMap[token].length;
    const suffixedToken = `${token}#${tokenOccurrenceIndex + 1}`; // note: we use `#` since it does not need to be escaped in regexp and its makes sense for an index due to it meaning "number"
    tokenToSuffixedTokensMap[token].push(suffixedToken);
    return suffixedToken;
  });

  // now go through the sql and replace each occurrence of a token with its suffixed,ordered identity (e.g., `WHERE (:until is null OR message.created_at <= :until)` => `WHERE (:until#1 is null OR message.created_at <= :until#2)`)
  const suffixedTokenSql = Object.keys(tokenToSuffixedTokensMap).reduce(
    (sqlNow, thisToken) => {
      const suffixedTokens = tokenToSuffixedTokensMap[thisToken];
      let sqlToUpdate = sqlNow;
      suffixedTokens.forEach((suffixedToken) => {
        sqlToUpdate = sqlToUpdate.replace(
          new RegExp(`(?<=[^:])(${thisToken})(?!#)`),
          suffixedToken,
        );
      });
      return sqlToUpdate;
    },
    sql,
  );

  // flatten the sql, to avoid subquery syntax interfering with below search patterns
  const strippedSql = strip(suffixedTokenSql);
  const { flattenedSql, references: subqueries } =
    flattenSqlByReferencingAndTokenizingSubqueries({
      sql: strippedSql,
    });

  // define the type definition for each variable
  const definitionsWithSuffixedTokens = suffixedInputVariableTokens.map(
    (token) => {
      // check if the token is actually in a subquery; run it on the subquery if so
      const foundSubqueryContainingToken = subqueries.find((subquery) =>
        new RegExp(`(?:[^\\w:])${token}`).test(subquery.sql),
      );
      if (foundSubqueryContainingToken)
        return extractTypeDefinitionFromInputVariableSql({
          token,
          sql: foundSubqueryContainingToken.sql,
        });

      // otherwise, run it on the root flattened query (flattened to make sure subquery syntax does not cause problems)
      return extractTypeDefinitionFromInputVariableSql({
        token,
        sql: flattenedSql,
      });
    },
  );

  // now cast the suffixed tokens defined in the type definitions back to the original tokens
  const suffixedTokenToTokenMap: {
    [index: string]: string;
  } = Object.fromEntries(
    Object.entries(tokenToSuffixedTokensMap)
      .map(([token, suffixedTokens]) => {
        return suffixedTokens.map((suffixedToken) => [suffixedToken, token]);
      })
      .flat(),
  );
  const definitions = definitionsWithSuffixedTokens.map(
    (definitionWithSuffixedToken) => {
      const tokenForSuffixedToken = suffixedTokenToTokenMap[
        `:${definitionWithSuffixedToken.name}`
      ].replace(':', '');
      return new TypeDefinitionOfQueryInputVariable({
        ...definitionWithSuffixedToken,
        name: tokenForSuffixedToken,
      });
    },
  );

  // return defs
  return definitions;
};
