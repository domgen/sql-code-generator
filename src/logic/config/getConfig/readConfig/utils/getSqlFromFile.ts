import { readFileAsync } from '../../../_utils/readFileAsync';

export const getSqlFromFile = async ({ filePath }: { filePath: string }): Promise<string> => {
  const extension = filePath.split('.').slice(-1)[0];
  if (extension === 'sql') return readFileAsync({ filePath });
  if (extension === 'ts') {
    const exports = await import(filePath);
    if (!exports.sql) throw new Error('.ts file did not have an export with name "sql"');
    return exports.sql;
  }
  throw new Error('path must lead to file with extension of .ts or .sql');
};
