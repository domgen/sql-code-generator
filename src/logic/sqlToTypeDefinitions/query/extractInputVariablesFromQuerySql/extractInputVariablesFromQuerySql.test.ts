import { extractInputVariablesFromQuerySql } from './extractInputVariablesFromQuerySql';
import { getSqlFromFile } from '../../../config/_utils/getSqlFromFile';

describe('extractSelectExpressionsFromQuerySql', () => {
  it('should be able to determine types accurately for this example', async () => {
    const sql = await getSqlFromFile({ filePath: `${__dirname}/../__test_assets__/find_image_by_id.sql` });
    const defs = extractInputVariablesFromQuerySql({ sql });
    expect(defs.length).toEqual(1);
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this other example', async () => {
    const sql = await getSqlFromFile({ filePath: `${__dirname}/../__test_assets__/select_suggestion.sql` });
    const defs = extractInputVariablesFromQuerySql({ sql });
    expect(defs.length).toEqual(0);
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this other other example', async () => {
    const sql = await getSqlFromFile({ filePath: `${__dirname}/../__test_assets__/upsert_suggestion.sql` });
    const defs = extractInputVariablesFromQuerySql({ sql });
    expect(defs.length).toEqual(5);
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for yet another example', async () => {
    const sql = await getSqlFromFile({
      filePath: `${__dirname}/../__test_assets__/find_all_suggestions_by_normalized_source.sql`,
    });
    const defs = extractInputVariablesFromQuerySql({ sql });
    expect(defs.length).toEqual(1);
    expect(defs).toMatchSnapshot();
  });
});
