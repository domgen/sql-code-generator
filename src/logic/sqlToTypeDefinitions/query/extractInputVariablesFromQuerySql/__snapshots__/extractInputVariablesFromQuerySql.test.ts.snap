// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for a case where an input variable is defined in a subquery used as an input to a function 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "serviceProviderUuid",
    "type": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "p.uuid",
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "bannerImageUuid",
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_provider_profile.input.1",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "pictureImageUuid",
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_provider_profile.input.2",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "introduction",
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_provider_profile.input.3",
      "tableReferencePath": null,
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for this example 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "id",
    "type": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "i.id",
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for this other example 1`] = `[]`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for this other other example 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "suggestionSource",
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_suggestion.input.0",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "externalId",
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_suggestion.input.1",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "suggestedIdeaId",
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_suggestion.input.2",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "status",
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_suggestion.input.3",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "result",
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_suggestion.input.4",
      "tableReferencePath": null,
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for yet another example 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "name",
    "type": TypeDefinitionReference {
      "functionReferencePath": "get_id_from_suggestion_source_name.input.0",
      "tableReferencePath": null,
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to extract input variables from a function that compares a timestamp with less-than-or-equals-to and null 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "threadUuid",
    "type": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "chat_thread.uuid",
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "until",
    "type": [
      "null",
    ],
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "until",
    "type": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "chat_message.created_at",
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "limit",
    "type": [
      "number",
    ],
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to extract input variables from a function that postgres::typecasts some of the input variables 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "homeStationGeocodeId",
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_train.input.0",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "combinationId",
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_train.input.1",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "locomotiveUuids",
    "type": TypeDefinitionReference {
      "functionReferencePath": "unnest.input.0",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "carriageUuids",
    "type": TypeDefinitionReference {
      "functionReferencePath": "unnest.input.0",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "engineerUuids",
    "type": TypeDefinitionReference {
      "functionReferencePath": "unnest.input.0",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "leadEngineerUuid",
    "type": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "train_engineer.uuid",
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "status",
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_train.input.6",
      "tableReferencePath": null,
    },
  },
]
`;
